# 1. Business Understanding
### 1.1 Background

The stock market plays a crucial role in economic development by enabling companies to raise capital and providing investors with opportunities to grow their wealth. In Kenya, the Nairobi Securities Exchange serves as the primary platform where publicly listed companies trade their shares.

Investors often rely on historical price trends, financial analysis, and market news when making investment decisions. However, predicting stock price movements remains a complex task due to market volatility, investor behavior, and external economic influences.

One of the most actively traded companies on the exchange is Safaricom. Due to its large market capitalization and high trading volume, Safaricom plays a significant role in influencing overall market activity. Analyzing and forecasting its stock price movements can therefore provide useful insights into investment strategies and market dynamics.

With the availability of historical market data and advancements in machine learning techniques, it is now possible to build predictive models that can identify patterns in stock price movements. These models can assist investors and analysts in making more informed decisions by estimating future price trends based on historical data.




### 1.2 Market Research

This section explains:

- how the Nairobi Securities Exchange works
- importance of stock prediction
- importance of Safaricom in the market
- existing prediction methods

The `Nairobi Securities Exchange`(NSE) is the primary securities exchange in Kenya where publicly listed companies trade shares and other financial instruments. The exchange plays a significant role in facilitating investment, capital formation, and economic growth within the country.

Stock prices in financial markets are influenced by a variety of factors including company performance, macroeconomic conditions, investor sentiment, and trading activity. As a result, forecasting stock price movements has become an important area of research in finance and data science.

Among the companies listed on the exchange, `Safaricom` is one of the most actively traded stocks. Due to its large market capitalization and high liquidity, Safaricom frequently accounts for a significant portion of total trading activity in the market. This makes it an important case study for analyzing stock price behavior in the Kenyan market.

In recent years, machine learning and data-driven approaches have been increasingly applied to financial forecasting. Techniques such as regression models, tree-based algorithms, and neural networks have shown potential in identifying hidden patterns within historical financial data. These methods can complement traditional financial analysis by providing predictive insights based on historical trading patterns.

By leveraging historical trading data from the Nairobi Securities Exchange, this project explores how machine learning techniques can be applied to predict stock prices and identify potential investment opportunities within the Kenyan market.

### 1.3. Problem Statement

Predicting stock price movements remains a significant challenge due to the dynamic and unpredictable nature of financial markets. Investors in the Nairobi Securities Exchange often rely on historical price trends and manual analysis when making investment decisions, which may not always capture hidden patterns within large datasets.

Despite the availability of historical trading data, there is limited use of data-driven predictive models specifically focused on stocks listed in the Kenyan market. This creates an opportunity to explore how machine learning techniques can be used to analyze historical stock data and generate reliable price forecasts.

This project seeks to develop a predictive model capable of estimating the next-day closing price of Safaricom using historical trading data. Additionally, the study will extend the analysis to a broader market perspective by applying the model to multiple stocks listed on the Nairobi Securities Exchange in order to identify the top 20 stocks with the highest predicted returns.


### 1.4. Objectives
#### Main Objective

To develop a machine learning model that predicts future stock prices using historical trading data from the Nairobi Securities Exchange.

Specific Objectives

1. To analyze historical trading data of Safaricom stock from 2007 to 2025.
2. To perform exploratory data analysis in order to understand price trends, volatility, and trading patterns..
3. To train and evaluate multiple machine learning models for predicting next-day stock prices.
4. To extend the prediction model to multiple stocks listed on the Nairobi Securities Exchange.
5. To rank stocks based on predicted returns and identify the top 20 potentially high-performing stocks.


### 1.5. Stakeholders

Several stakeholders may benefit from insights generated in this project:

1. Investors and Traders
- Individual and institutional investors can use predictive insights to support investment decision-making and identify potential opportunities within the stock market.

2. Financial Analysts
- Market analysts can utilize predictive models to complement traditional financial analysis methods when evaluating stock performance.

3. tock Brokerage Firms
- Brokerage firms may use predictive insights to provide data-driven recommendations to their clients.

4. Researchers and Data Scientists
- This project contributes to the growing field of financial data science by demonstrating how machine learning techniques can be applied to stock market prediction using local market data.

5. The Nairobi Securities Exchange
- Insights derived from this analysis may contribute to a deeper understanding of trading patterns and market dynamics within the exchange.

# 2. Data Understanding
### 2.1 Data Sources

The dataset used in this project consists of historical stock market trading data from the Nairobi Securities Exchange covering the period `2007` to `OCT 2025`. The data contains daily trading records for multiple companies listed on the exchange.

Each record includes information such as the trading date, stock code, company name, daily price movements, trading volume, and other related indicators.

The dataset was collected in multiple files organized by year. To facilitate analysis, the files are combined into a single dataset that contains all trading records within the study period.

*importing the required libraries and loading the data* 

In [1]:
import os
import glob
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path

In [2]:
# Define the data path
DATA_PATH = Path("../data/raw_data")

print(DATA_PATH)

..\data\raw_data


First, we load a sample file to inspect the data. 

In [3]:
# Load the 2007 data into a variable
df_2007 = pd.read_csv(DATA_PATH / "NSE_data_all_stocks_2007.csv")

# Inspect the first 5 rows
df_2007.head()

,DATE,CODE,NAME,12m Low,12m High,Day Low,Day High,Day Price,Previous,Change,Change%,Volume,Adjust
0,1/2/2007,EGAD,Eaagads,22,57,52,52,52,52,-,-,300,26
1,1/2/2007,KAPC,Kapchorua Tea,111,148,100,100,100,98.5,1.5,1.50%,"7,800",-
2,1/2/2007,KUKZ,Kakuzi,67.5,89,42.5,45,43.5,42.25,1.25,2.87%,"2,700",-
3,1/2/2007,REA,Rea Vipingo Plantations,14.5,23.5,25,26.5,25.5,25.75,-0.25,0.98%,"10,500",-
4,1/2/2007,SASN,Sasini Tea and Coffee,10.5,13.6,140,145,140,141,-1,0.71%,"70,900",23.33


In [4]:
# Inspect the datasets details (nulls, number of columns etc.)
df_2007.info()

<class 'pandas.DataFrame'>
RangeIndex: 10608 entries, 0 to 10607
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   DATE       10608 non-null  str  
 1   CODE       10608 non-null  str  
 2   NAME       10585 non-null  str  
 3   12m Low    10608 non-null  str  
 4   12m High   10608 non-null  str  
 5   Day Low    10608 non-null  str  
 6   Day High   10608 non-null  str  
 7   Day Price  10608 non-null  str  
 8   Previous   10608 non-null  str  
 9   Change     10608 non-null  str  
 10  Change%    10608 non-null  str  
 11  Volume     10608 non-null  str  
 12  Adjust     10608 non-null  str  
dtypes: str(13)
memory usage: 1.1 MB


In [5]:
# Print columns
df_2007.columns

Index(['DATE', 'CODE', 'NAME', '12m Low', '12m High', 'Day Low', 'Day High',
       'Day Price', 'Previous', 'Change', 'Change%', 'Volume', 'Adjust'],
      dtype='str')

### Combine All Datasets
As mentioned ealier, we are combining the datasets under one variable for easy access 

In [6]:
# Find all files in the raw_data folder
files = glob.glob("../data/raw_data/*.csv")

# Ensure all files are found
files

['../data/raw_data\\NSE_data_all_stocks_2007.csv',
 '../data/raw_data\\NSE_data_all_stocks_2008.csv',
 '../data/raw_data\\NSE_data_all_stocks_2009.csv',
 '../data/raw_data\\NSE_data_all_stocks_2010.csv',
 '../data/raw_data\\NSE_data_all_stocks_2011.csv',
 '../data/raw_data\\NSE_data_all_stocks_2012.csv',
 '../data/raw_data\\NSE_data_all_stocks_2013.csv',
 '../data/raw_data\\NSE_data_all_stocks_2014.csv',
 '../data/raw_data\\NSE_data_all_stocks_2015.csv',
 '../data/raw_data\\NSE_data_all_stocks_2016.csv',
 '../data/raw_data\\NSE_data_all_stocks_2017.csv',
 '../data/raw_data\\NSE_data_all_stocks_2018.csv',
 '../data/raw_data\\NSE_data_all_stocks_2019.csv',
 '../data/raw_data\\NSE_data_all_stocks_2020.csv',
 '../data/raw_data\\NSE_data_all_stocks_2021.csv',
 '../data/raw_data\\NSE_data_all_stocks_2022.csv',
 '../data/raw_data\\NSE_data_all_stocks_2023.csv',
 '../data/raw_data\\NSE_data_all_stocks_2024.csv',
 '../data/raw_data\\NSE_data_all_stocks_2025.csv']

In [7]:
# Create an empty list
dfs = []

# Loop over every file and read the contents
for file in files:
    df = pd.read_csv(file)

    # Append the dataframes to the empty list, dfs, to create a list of dataframes with no index
    dfs.append(df)

# Confirm the code worked
dfs

[             DATE   CODE                     NAME   12m Low  12m High  \
 0        1/2/2007   EGAD                  Eaagads        22        57   
 1        1/2/2007   KAPC            Kapchorua Tea       111       148   
 2        1/2/2007   KUKZ                   Kakuzi      67.5        89   
 3        1/2/2007    REA  Rea Vipingo Plantations      14.5      23.5   
 4        1/2/2007   SASN    Sasini Tea and Coffee      10.5      13.6   
 ...           ...    ...                      ...       ...       ...   
 10603  12/24/2007   EVRD     Eveready East Africa       1.4         3   
 10604  12/24/2007    MSC             Mumias Sugar         4      7.35   
 10605  12/24/2007   UNGA               Unga Group       9.6      16.5   
 10606  12/24/2007   ACCS        AccessKenya Group       3.9      10.1   
 10607  12/24/2007  ^N20I       NSE 20-Share Index  3,396.83  5,030.91   
 
         Day Low  Day High Day Price  Previous  Change Change%      Volume  \
 0            52        52      

In [8]:
# Concatenate the dataframes to create one dataframe
combined_df = pd.concat(dfs, ignore_index=True)

# Review the first 15 rows
combined_df.head(15)

,DATE,CODE,NAME,12m Low,12m High,Day Low,Day High,Day Price,Previous,Change,Change%,Volume,Adjust,Date,Adjusted,Code,Name,Adjusted Price
0,1/2/2007,EGAD,Eaagads,22,57,52,52,52,52,-,-,300,26,NaN,NaN,NaN,NaN,NaN
1,1/2/2007,KAPC,Kapchorua Tea,111,148,100,100,100,98.5,1.5,1.50%,"7,800",-,NaN,NaN,NaN,NaN,NaN
2,1/2/2007,KUKZ,Kakuzi,67.5,89,42.5,45,43.5,42.25,1.25,2.87%,"2,700",-,NaN,NaN,NaN,NaN,NaN
3,1/2/2007,REA,Rea Vipingo Plantations,14.5,23.5,25,26.5,25.5,25.75,-0.25,0.98%,"10,500",-,NaN,NaN,NaN,NaN,NaN
4,1/2/2007,SASN,Sasini Tea and Coffee,10.5,13.6,140,145,140,141,-1,0.71%,"70,900",23.33,NaN,NaN,NaN,NaN,NaN
5,1/2/2007,UTK,Unilever Tea Kenya,45,45,80,82,81,81.5,-0.5,0.62%,"1,000",-,NaN,NaN,NaN,NaN,NaN
6,1/2/2007,WTK,Williamson Tea Kenya,180,290,120,120,120,118,2,1.67%,"10,000",-,NaN,NaN,NaN,NaN,NaN
7,1/2/2007,C&G,Car and General (K),21,29,50,50,50,50,-,-,"66,700",33.33,NaN,NaN,NaN,NaN,NaN
8,1/2/2007,CMC,CMC Holdings,13.5,13.5,173,193,188,173,15,7.98%,"36,200",15.67,NaN,NaN,NaN,NaN,NaN
9,1/2/2007,FIRE,Sameer Africa,3.4,5.95,22.5,26,24.75,24,0.75,3.03%,"54,800",-,NaN,NaN,NaN,NaN,NaN


In [9]:
# Review the last 15 rows
combined_df.tail(15)

,DATE,CODE,NAME,12m Low,12m High,Day Low,Day High,Day Price,Previous,Change,Change%,Volume,Adjust,Date,Adjusted,Code,Name,Adjusted Price
300990,NaN,NaN,NaN,0.27,0.27,0.27,0.27,0.27,0.27,-,-,-,NaN,31-Oct-25,NaN,MSC,Mumias Sugar Company Ltd,-
300991,NaN,NaN,NaN,5.8,9.0,7.28,8.0,7.46,8.06,-0.6,-7.44%,6820,NaN,31-Oct-25,NaN,SKL,Shri Krishana Overseas Plc,-
300992,NaN,NaN,NaN,14.4,28.25,24.5,25.0,24.65,24.7,-0.05,-0.20%,1569,NaN,31-Oct-25,NaN,UNGA,Unga Group Ltd,-
300993,NaN,NaN,NaN,14.1,31.0,29.6,30.7,30.25,29.55,0.7,2.37%,3051822,NaN,31-Oct-25,NaN,SCOM,Safaricom Plc,-
300994,NaN,NaN,NaN,20.0,20.0,20.0,20.0,20.0,20,-,-,-,NaN,31-Oct-25,NaN,LAPR,Laptrust Imara Income-REIT,-
300995,NaN,NaN,NaN,3000.0,9800.0,4840.0,5175.0,4910.0,4835,75,1.55%,539,NaN,31-Oct-25,NaN,GLD,ABSA NewGold ETF,-
300996,NaN,NaN,NaN,755.0,1000.0,854.0,857.0,855.0,857,-2,-0.23%,27,NaN,31-Oct-25,NaN,SMWF,Satrix MSCI World Feeder ETF,-
300997,NaN,NaN,NaN,1170.1,1934.23,1910.77,1910.77,1910.77,1877.52,33.25,1.77%,-,NaN,31-Oct-25,NaN,^N10I,NSE 10-Share Index,-
300998,NaN,NaN,NaN,1823.17,3153.39,3116.69,3116.69,3116.69,3090.7,25.99,0.84%,-,NaN,31-Oct-25,NaN,^N20I,NSE 20-Share Index,-
300999,NaN,NaN,NaN,3063.32,5052.98,4998.39,4998.39,4998.39,4900.89,97.5,1.99%,-,NaN,31-Oct-25,NaN,^N25I,NSE 25-Share Index,-


In [10]:
# Turn it into a CSV file
combined_df.to_csv("../data/raw_data/processed/NSE_data_all_stocks_2007-2025.csv", index=False)

#### Inspect the merged data

In [11]:
# Shape of the dataset
print("Rows, Columns:", combined_df.shape)

Rows, Columns: (301005, 18)


In [12]:
combined_df.columns

Index(['DATE', 'CODE', 'NAME', '12m Low', '12m High', 'Day Low', 'Day High',
       'Day Price', 'Previous', 'Change', 'Change%', 'Volume', 'Adjust',
       'Date', 'Adjusted', 'Code', 'Name', 'Adjusted Price'],
      dtype='str')

In [13]:
# Data types and non-null values
combined_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 301005 entries, 0 to 301004
Data columns (total 18 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   DATE            223107 non-null  str   
 1   CODE            233690 non-null  str   
 2   NAME            233642 non-null  str   
 3   12m Low         301003 non-null  object
 4   12m High        301003 non-null  object
 5   Day Low         301003 non-null  object
 6   Day High        301003 non-null  object
 7   Day Price       301003 non-null  object
 8   Previous        301003 non-null  str   
 9   Change          301003 non-null  str   
 10  Change%         301003 non-null  str   
 11  Volume          301003 non-null  str   
 12  Adjust          223107 non-null  str   
 13  Date            77896 non-null   str   
 14  Adjusted        10583 non-null   str   
 15  Code            67313 non-null   str   
 16  Name            67313 non-null   str   
 17  Adjusted Price  67313 non-null   str   


In [14]:
# Summary stats for numerical columns
combined_df.describe()

,DATE,CODE,NAME,12m Low,12m High,Day Low,Day High,Day Price,Previous,Change,Change%,Volume,Adjust,Date,Adjusted,Code,Name,Adjusted Price
count,223107,233690,233642,301003,301003,301003,301003,301003,301003,301003,301003,301003,223107,77896,10583,67313,67313,67313
unique,3491,99,143,903,987,23058,23058,23967,21596,5960,2353,31270,3905,1195,1063,79,81,45
top,24-Mar-17,HFCK,NSE 20-Share Index,4,55.5,5.5,5.5,5.5,5.5,-,-,-,-,6-Oct-25,-,EGAD,Eaagads Ltd,-
freq,138,3744,3645,5689,3848,2968,2909,2721,2738,111261,111368,66398,189965,75,6262,946,946,67254


In [15]:
# Check unique values for each column (useful for categorical vars)
combined_df.nunique()

DATE               3491
CODE                 99
NAME                143
12m Low             903
12m High            987
Day Low           23058
Day High          23058
Day Price         23967
Previous          21596
Change             5960
Change%            2353
Volume            31270
Adjust             3905
Date               1195
Adjusted           1063
Code                 79
Name                 81
Adjusted Price       45
dtype: int64

In [16]:
# Total and percent missing
missing = combined_df.isnull().sum()
missing_percent = (missing / len(combined_df)) * 100
missing_df = pd.DataFrame({'Missing Values': missing, 'Percent': missing_percent})
missing_df[missing_df['Missing Values'] > 0]

,Missing Values,Percent
DATE,77898,25.879304
CODE,67315,22.363416
NAME,67363,22.379362
12m Low,2,0.000664
12m High,2,0.000664
Day Low,2,0.000664
Day High,2,0.000664
Day Price,2,0.000664
Previous,2,0.000664
Change,2,0.000664


In [17]:
# Check duplicates
combined_df.duplicated().sum()

np.int64(89)

## Data Cleaning

Data cleaning is a critical step in the data science workflow that involves identifying and correcting errors, inconsistencies, and inaccuracies in a dataset. Raw data collected from different sources or across long time periods often contains issues such as missing values, inconsistent column naming, duplicate variables, incorrect data formats, and redundant information. If these issues are not addressed, they can negatively affect the quality of analysis and the performance of predictive models.

In the context of this project, the Nairobi Securities Exchange (NSE) dataset was compiled from multiple yearly records spanning from 2007 to 2025. Because the data originates from different time periods and potentially different reporting formats, several inconsistencies were observed during the data understanding phase. These include variations in column naming conventions (for example, `DATE` vs `Date`), duplicate columns representing the same variable, automatically generated index columns, and missing values resulting from the merging process.

The objective of the data cleaning stage is therefore to transform the raw merged dataset into a structured and consistent format suitable for analysis and modeling. The key steps performed in this phase include:

- Resolving inconsistencies in column names.
- Merging duplicate columns representing the same variables.
- Removing redundant or unnecessary columns.
- Standardizing the structure and ordering of dataset columns.
- Preparing the dataset for further preprocessing steps such as data type conversion and exploratory data analysis.

By completing these cleaning steps, the dataset becomes more reliable and suitable for downstream tasks including exploratory analysis, feature engineering, and the development of machine learning models for stock price prediction.

### Handling Column Name Inconsistencies

After merging the NSE datasets from multiple years, several duplicated columns were identified. These duplicates occurred because different datasets used different capitalization styles for the same variables.

Examples include:

| Column Version 1 | Column Version 2 |
|------------------|------------------|
| DATE | Date |
| CODE | Code |
| NAME | Name |
| Adjust | Adjusted |
| Adjusted Price | Adjust |

Because pandas treats column names as case-sensitive, these columns were interpreted as separate variables during the merge process. As a result, some rows contain values in one column while the corresponding duplicate column contains missing values.

### Cleaning Strategy

To resolve this issue:

1. Merge duplicated columns representing the same variable.
2. Fill missing values in one column using values from the other column.
3. Retain a single standardized column name.
4. Remove redundant columns after merging.

This ensures the dataset maintains a consistent schema across all observations.

In [18]:
combined_df.columns

Index(['DATE', 'CODE', 'NAME', '12m Low', '12m High', 'Day Low', 'Day High',
       'Day Price', 'Previous', 'Change', 'Change%', 'Volume', 'Adjust',
       'Date', 'Adjusted', 'Code', 'Name', 'Adjusted Price'],
      dtype='str')

In [19]:
# combine Date and DATE columns into one column called date
combined_df['date'] = combined_df['DATE'].fillna(combined_df['Date'])

In [20]:
# check the new date column for nulls
combined_df['date'].isna().sum()

np.int64(2)

In [21]:
# combine Code and CODE columds
combined_df['code'] = combined_df['CODE'].fillna(combined_df['Code'])

In [22]:
#Check the code colums for nulls
combined_df['code'].isna().sum()

np.int64(2)

In [23]:
# combine name and NAME columns
combined_df['name'] = combined_df['NAME'].fillna(combined_df['Name'])

In [24]:
# check the name column for nulls
combined_df['name'].isna().sum()

np.int64(50)

In [25]:
# combine Adjusted Price, Adjust and Adjusted columns into one column called adjusted_price
combined_df['adjusted_price'] = combined_df['Adjusted Price'].fillna(combined_df['Adjust'])
combined_df['adjusted_price'] = combined_df['adjusted_price'].fillna(combined_df['Adjusted'])

In [26]:
combined_df['adjusted_price'].isna().sum()

np.int64(2)

In [27]:
combined_df.columns

Index(['DATE', 'CODE', 'NAME', '12m Low', '12m High', 'Day Low', 'Day High',
       'Day Price', 'Previous', 'Change', 'Change%', 'Volume', 'Adjust',
       'Date', 'Adjusted', 'Code', 'Name', 'Adjusted Price', 'date', 'code',
       'name', 'adjusted_price'],
      dtype='str')

In [28]:
# drop the original columns that have been combined into new columns
combined_df = combined_df.drop(columns=[
'DATE','Date',
'CODE','Code',
'NAME','Name',
'Adjust','Adjusted','Adjusted Price'
])

In [29]:
combined_df.columns

Index(['12m Low', '12m High', 'Day Low', 'Day High', 'Day Price', 'Previous',
       'Change', 'Change%', 'Volume', 'date', 'code', 'name',
       'adjusted_price'],
      dtype='str')

In [30]:
combined_df.head()

,12m Low,12m High,Day Low,Day High,Day Price,Previous,Change,Change%,Volume,date,code,name,adjusted_price
0,22,57,52,52,52,52,-,-,300,1/2/2007,EGAD,Eaagads,26
1,111,148,100,100,100,98.5,1.5,1.50%,"7,800",1/2/2007,KAPC,Kapchorua Tea,-
2,67.5,89,42.5,45,43.5,42.25,1.25,2.87%,"2,700",1/2/2007,KUKZ,Kakuzi,-
3,14.5,23.5,25,26.5,25.5,25.75,-0.25,0.98%,"10,500",1/2/2007,REA,Rea Vipingo Plantations,-
4,10.5,13.6,140,145,140,141,-1,0.71%,"70,900",1/2/2007,SASN,Sasini Tea and Coffee,23.33


## Column Cleanup and Standardization

After merging and consolidating duplicate columns, the dataset still contained unnecessary index columns (`Unnamed: 0` and `Unnamed: 0.1`). These columns were automatically generated during previous CSV exports and do not contain meaningful analytical information.

To prepare the dataset for analysis, the following steps were performed:

1. Remove redundant index columns (`Unnamed: 0`, `Unnamed: 0.1`).
2. Standardize column names to a consistent naming format.
3. Reorder the columns to match the desired schema for analysis.

The final dataset structure will be:

Date, Code, Name, 12m Low, 12m High, Day Low, Day High, Day Price, Previous, Change, Change%, Volume, Adjusted Price

This structure ensures clarity and consistency before proceeding to data type corrections and exploratory analysis.

In [31]:
combined_df = combined_df.drop(columns=['Unnamed: 0', 'Unnamed: 0.1'], errors='ignore')

In [32]:
# standardize column names to lowercase and replace spaces with underscores
combined_df.columns = combined_df.columns.str.lower().str.replace(' ', '_')

In [33]:
combined_df.columns

Index(['12m_low', '12m_high', 'day_low', 'day_high', 'day_price', 'previous',
       'change', 'change%', 'volume', 'date', 'code', 'name',
       'adjusted_price'],
      dtype='str')

In [34]:
combined_df = combined_df[[
'date',
'code',
'name',
'12m_low',
'12m_high',
'day_low',
'day_high',
'day_price',
'previous',
'change',
'change%',
'volume',
'adjusted_price'
]]

combined_df.head()

,date,code,name,12m_low,12m_high,day_low,day_high,day_price,previous,change,change%,volume,adjusted_price
0,1/2/2007,EGAD,Eaagads,22,57,52,52,52,52,-,-,300,26
1,1/2/2007,KAPC,Kapchorua Tea,111,148,100,100,100,98.5,1.5,1.50%,"7,800",-
2,1/2/2007,KUKZ,Kakuzi,67.5,89,42.5,45,43.5,42.25,1.25,2.87%,"2,700",-
3,1/2/2007,REA,Rea Vipingo Plantations,14.5,23.5,25,26.5,25.5,25.75,-0.25,0.98%,"10,500",-
4,1/2/2007,SASN,Sasini Tea and Coffee,10.5,13.6,140,145,140,141,-1,0.71%,"70,900",23.33


check for missing data after the merge

In [35]:
# Total and percent missing
missing_merged = combined_df.isnull().sum()
missing_percent = (missing_merged / len(combined_df)) * 100
missing_df = pd.DataFrame({'Missing Values': missing_merged, 'Percent': missing_percent})
missing_df[missing_df['Missing Values'] > 0]

,Missing Values,Percent
date,2,0.000664
code,2,0.000664
name,50,0.016611
12m_low,2,0.000664
12m_high,2,0.000664
day_low,2,0.000664
day_high,2,0.000664
day_price,2,0.000664
previous,2,0.000664
change,2,0.000664


In [38]:
# generate a copy of a final dataset
combined_df.to_csv("../data/raw_data/processed/nse_final_dataset.csv", index=False)

In [39]:
combined_df.describe()

,date,code,name,12m_low,12m_high,day_low,day_high,day_price,previous,change,change%,volume,adjusted_price
count,301003,301003,300955,301003,301003,301003,301003,301003,301003,301003,301003,301003,301003
unique,4686,107,154,903,987,23058,23058,23967,21596,5960,2353,31270,4031
top,24-Mar-17,HFCK,NSE 20-Share Index,4,55.5,5.5,5.5,5.5,5.5,-,-,-,-
freq,138,4690,4486,5689,3848,2968,2909,2721,2738,111261,111368,66398,263481


#### Data Type Correction

After completing the initial data cleaning process and producing a consolidated dataset, the next step involves correcting the data types of several variables. When datasets are imported from CSV files, columns containing numeric values may sometimes be interpreted as text due to the presence of formatting characters such as commas or percentage signs.

For example, columns such as **Volume** may contain values formatted with commas (e.g., "7,800"), while **Change%** includes percentage symbols (e.g., "1.50%"). Additionally, the **Date** column is initially stored as a string rather than a datetime object.

These formatting issues prevent proper numerical analysis and must therefore be corrected. In this stage, the dataset will be prepared for analysis by:

- Converting the **Date** column to a datetime format.
- Removing commas from the **Volume** column and converting it to numeric.
- Removing percentage symbols from **Change%** and converting it to a numeric value.
- Ensuring all price-related columns are stored as numeric data types.

Correcting these data types ensures that the dataset can be used effectively for exploratory analysis and machine learning models.